File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.cc
#include <iostream>
#include <algorithm>
#include <vector>
#define ENDL '\n'

using namespace std;
using ll = long long;


int main(void) {
	ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);
	int K, N; cin >> K >> N;
	vector<int>lines(K, 0);
	for (int i = 0; i < K; i++) cin >> lines[i];
	sort(lines.begin(), lines.end());
	ll minRange = 1;
	ll maxRange = lines.back();
	int answer = 0;
	while (minRange <= maxRange) {
		ll mid = (minRange + maxRange) / 2;
		int sum = 0;
		for (int len : lines) {
			sum += len / mid;
		}
		if (sum >= N) {
			answer = mid;
			minRange = mid + 1;
		}
		else {
			maxRange = mid - 1;
		}
	}
	cout << answer;

	return 0;
}

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int have, need;

        StringTokenizer st = new StringTokenizer(br.readLine());
        have = Integer.parseInt(st.nextToken());
        need = Integer.parseInt(st.nextToken());

        int[] lines = new int[have];

        for(int i = 0; i < have; i++) {
            lines[i] = Integer.parseInt(br.readLine());
        }

        Arrays.sort(lines);

        long minRange = 1;
        long maxRange = lines[have-1];

        long answer = Integer.MIN_VALUE;
        while(minRange <= maxRange) {
            long mid = (minRange + maxRange) / 2;

            long tmp = 0;
            for(int l : lines) {
                tmp += l / mid;
            }

            if(tmp >= need) {
                minRange = mid + 1;
                answer = Math.max(answer, mid);
            } else {
                maxRange = mid - 1;
            }
        }

        System.out.println(answer);

    }
}